### PR TITLE
Add preferences to configure audio codecs and 4k transcoding/direct play

### LINKED ISSUE #2867: Add a fallback to AC3/EAC3 option
### Describe the feature you'd like

Some audio codecs such as TrueHD and DTS dont always play correctly so I think the easiest way to get bitstream surround sound is to request a AC3 or EAC3 transcode from the server instead of an AAC or any other codec.

From what I believe bitstreaming AC3 or EAC3 is the only way to get surround over optical so this would be a great addition

### PR DESCRIPTION
**Pull request by t0mas**
Allow users to select which audio codecs to direct play, and in case of transcoding select the preferred codec to transcode into. Main use-case is for systems where the connected audio equipment fails to play some codecs. For example transcode Dolby TrueHD and DTS into AC3 or EAC3 if those are supported by the audio receiver or select AAC if more advanced formats are not supported.

Allow users to select whether 4k video streams can be accepted or not. Main example is the "Chromecast with Google TV" device, which in the old method will request 4k streams but is unable to play those (there exists a 4K version of the device). The new preference allows users to disable 4k if their player is not capable of playing it. The setting defaults to the same as the original behaviour (device detection).

I saw PR #3109 by @MichaelRUSF overlaps with this but only does AAC to DolbyDigital and not the other options.

**Changes**
- Add user preferences to enable audio codecs
- Add user preference to select transcode audio target
- Add user preference to enable/disable 4k video
- Changed ExoPlayerProfile to be based on the selected preferences

**Issues**
Fixes #2867, #2602, #2991 for audio and #2971, #2516 for video

### CODE REVIEW

**Comment by t0mas** [comment]
Looking at older issues, this also fixes #2432, #1765, #618

Could be useful to keep all those issue numbers together when either this PR or another solution is merged to close a lot of older ones.

**Comment by nielsvanvelzen** [comment]
I'm currently focusing on getting 0.16 out and bugfixes can still be merged but larger changes (especially those with string changes) are unfortunately going to wait for 0.17. This also means it will take some time before I'll review the pull request.

Additionally, my focus with 0.17 is to implement the new playback code for all video playback and completely removing the old implementation which will basically remove all changes from this PR again. The profiles will work quite differently in the rewrite (with fancy auto detection and such) so it won't be easy to port these changes over. Individual toggles for all codecs/formats are planned though.

**Comment by SplitMilky** [comment]
I think #520 would be fixed as well.

**Review by saxobroko — APPROVED** [review]

File: app/src/main/java/org/jellyfin/androidtv/preference/UserPreferences.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/preference/UserPreferences.kt:107]: Preference keys should not start with `pref_`. This is only used on old keys for compatibility reasons (creating a migration just for a rename of the key is kinda pointless)

File: app/src/main/java/org/jellyfin/androidtv/preference/UserPreferences.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/preference/UserPreferences.kt:153]: I don't think we should shorten this to `thd`, doesn't seem like an "official" abbreviation.

File: app/src/main/java/org/jellyfin/androidtv/util/profile/ExoPlayerProfile.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/util/profile/ExoPlayerProfile.kt:33]: This is not correctly formatted.

```suggestion
	audioDirectPlayCodecs: Array<String>,
	audioTranscodeTarget: Array<String>,
	enable4KSupport: Boolean
```

File: app/src/main/res/values/strings.xml
Comment (nielsvanvelzen) [inline app/src/main/res/values/strings.xml:180]:
```suggestion
    <string name="pref_audio_directplay">Direct play audio codecs</string>
```

File: app/src/main/java/org/jellyfin/androidtv/ui/playback/PlaybackController.java
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/playback/PlaybackController.java:637]:
```suggestion
                if (directPlayCodecs.length > 0) return directPlayCodecs;
```

File: app/src/main/java/org/jellyfin/androidtv/ui/preference/screen/PlaybackPreferencesScreen.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/preference/screen/PlaybackPreferencesScreen.kt:207]: I think it would be better if add a new preference screen for this category. You can look at `UserPreferencesScreen` to see how to link to one.

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]
Sorry for the long wait. I've done an initial review on the changes and got some small comments.

**Comment by zkhcohen** [comment]
This is a great addition to Jellyfin's config.

It would also serve as a workaround for: https://github.com/jellyfin/jellyfin-androidtv/issues/3227

**Comment by PriceChild** [comment]
This is awesome. I have a 1st gen fire TV and this makes it very easy to disable aac. Audio is being transcoded to my preferred format using the debug build I built myself from this branch.

Unfortunately (and identical to one of the other branches solution) although I get perfect 5.1 audio, the video takes seconds per frame, skipping.

Using the debug build, with aac disabled, If I check the server dashboard when playing a specific file, I see under reasons for transcode is that both audio and video are not supported by the player.

Using the debug build, with aac enabled, I get perfect video and stereo audio. The server dashboard explaining it is transcoding for video.

If I use the fire TV store's release build, I see under reasons for transcode only video being the reason.

So the server is transcoding in both cases, about 80fps reported by the server.

Is there any more troubleshooting I can provide to understand why taking advantage of this change helps, but also regresses my device's performance?

**Comment by anym001** [comment]
Where can I find these settings for audio codec directplay?

**Comment by PriceChild** [comment]
@anym001 This isn't in a released build. Have you installed a build based on this branch?

**Comment by anym001** [comment]
@PriceChild Oh okay, I wasn't aware of that. I am currently on the stable release.

**Comment by nielsvanvelzen** [comment]
> Is it already foreseeable when this build will be released?

The pull request is currently waiting on @t0mas to update it so I can re-review and potentially merge. It will then be included in 0.17 (no ETA). If no activity happens the pull request will eventually be closed as stale.

**Comment by t0mas** [comment]
@nielsvanvelzen Ok, thanks for reviewing! I'll look into the separate screen option. Quite busy at the moment so it may take a while.

**Comment by t0mas** [comment]
> Is there any more troubleshooting I can provide to understand why taking advantage of this change helps, but also regresses my device's performance?

This branch also adds a setting to enable and disable 4k video playback. The Fire stick is one of the devices that is explicitly listed in the original code to disable 4k playback, because it doesn't have enough performance to play it. Did you test with a 1080p video or with a 4k one?

File: app/src/main/java/org/jellyfin/androidtv/ui/playback/PlaybackController.java
Comment (MichaelRUSF) [inline app/src/main/java/org/jellyfin/androidtv/ui/playback/PlaybackController.java:624]: Since you’re adding a new panel and there are only two other codecs, I would add them as separate options.

File: app/src/main/java/org/jellyfin/androidtv/ui/preference/screen/PlaybackPreferencesScreen.kt
Comment (MichaelRUSF) [inline app/src/main/java/org/jellyfin/androidtv/ui/preference/screen/PlaybackPreferencesScreen.kt:261]: As commented above, I think it makes sense to split out OPUS and FLAC as separate options.

File: app/src/main/java/org/jellyfin/androidtv/ui/playback/PlaybackController.java
Comment (MichaelRUSF) [inline app/src/main/java/org/jellyfin/androidtv/ui/playback/PlaybackController.java:612]:
```suggestion
            codecs.add(Codec.Audio.MLP);
            codecs.add(Codec.Audio.TRUEHD);
```

File: app/src/main/java/org/jellyfin/androidtv/ui/playback/PlaybackController.java
Comment (MichaelRUSF) [inline app/src/main/java/org/jellyfin/androidtv/ui/playback/PlaybackController.java:622]:
```suggestion

```

**Comment by cgernert** [comment]
I have a Sony (Google) TV and am also trying to solve "AAC Plays as Stereo" https://github.com/jellyfin/jellyfin-androidtv/issues/2602

I have built 
https://github.com/t0mas/jellyfin-androidtv/tree/transcode-preferences 
and it does not work despite trying different configurations: Transcoding to PCM, DD, DTS. AAC checked and unchecked as supported format. I got 2.0 or 2.1 PCM (depending on TV audio settings) from the 5.1 AAC stream. No transcoding.

On the other hand, I built 
https://github.com/MichaelRUSF/jellyfin-androidtv/tree/master 
and it works. I get 5.1 PCM from the 5.1 AAC stream.

I do not know if this behavior is intended or not.

**Comment by nielsvanvelzen** [comment]
> I have built [t0mas/jellyfin-androidtv@`transcode-preferences`](https://github.com/t0mas/jellyfin-androidtv/tree/transcode-preferences?rgh-link-date=2024-02-20T16%3A14%3A19Z) and it does not work despite trying different configurations: Transcoding to PCM, DD, DTS. AAC checked and unchecked as supported format. I got 2.0 or 2.1 PCM (depending on TV audio settings) from the 5.1 AAC stream. No transcoding.
> 
> On the other hand, I built [MichaelRUSF/jellyfin-androidtv@`master`](https://github.com/MichaelRUSF/jellyfin-androidtv/tree/master?rgh-link-date=2024-02-20T16%3A14%3A19Z) and it works. I get 5.1 PCM from the 5.1 AAC stream.

You're comparing a pull request that is almost 300 commits behind with a fork that is only 11 behind. Not sure what you're getting at.

**Comment by SplitMilky** [comment]
So what are we waiting on here?  For nielsvanvelzen  comments to be incorporated in?  Can i do this directly or do I have to make a fork of this fork to add  suggested comment changes?

**Comment by cgernert** [comment]
> You're comparing a pull request that is almost 300 commits behind with a fork that is only 11 behind. Not sure what you're getting at.

So the code that was recently (2 weeks ago) reviewed by MichaelRUSF on https://github.com/jellyfin/jellyfin-androidtv/pull/3110/files/7e36ca6af03d11c4f1dc93bf4e97ccf3c435d2b5 is just the settings in the UI?
I just wonder why this code was recently reviewed, resulting in an incomplete(?) fix for the issue.

**Comment by nielsvanvelzen** [comment]
@t0mas have you been able to look at this PR again?

**Comment by dandud100** [comment]
> @t0mas have you been able to look at this PR again?

It's been a week without a reply from @t0mas, and I'd like to inquire about the status of this PR. Could someone else possibly take over to finalize it, or could you provide an update on how we'll proceed from here?

**Comment by nielsvanvelzen** [comment]
If the pull request becomes stale (e.g. author is unresponsive/unwilling to continue with the PR) it will eventually be closed. Someone else can of course take over (with permission from the author).

**Comment by t0mas** [comment]
@dandud100 I unfortunately don't have time at the moment, but of course no objections if someone else wants to make the required changes.

**Comment by dandud100** [comment]
> (with permission from the author)
t0mas don't have time at this moment but replied here to give the permission for someone to take this over.

@MichaelRUSF are you able to continue this pr?

**Comment by majmongoose** [comment]
If all this request needs is the changes specified in previous comments, I can continue some work on this PR.

**Comment by dandud100** [comment]
> If all this request needs is the changes specified in previous comments, I can continue some work on this PR.

Would be epic!

**Comment by nielsvanvelzen** [comment]
There would definitely be additional review feedback, so it's not just rebasing + applying the current feedback comments.

**Comment by nielsvanvelzen** [comment]
Closing as stale. Please re-open or create a new pull request if interested in finishing.

**Comment by dandud100** [comment]
@majmongoose are you able to work on it or not?

Btw. Is there an option to add an bounty to this pr?

**Comment by majmongoose** [comment]
> @majmongoose are you able to work on it or not?
> 
> Btw. Is there an option to add an bounty to this pr?

I did further review of the previous committers code and as Neils stated it would be a significant amount of work to bring it up to date; which is unfortunately beyond my skill set and availability. 

I did make a PR (since accepted) to add the Chromecast HD (CCHD) to the list of devices which is forced to a max of 1080p, but it's not in any of the app releases yet. However the server appears to be ignoring the limit and sends the CCHD 4k anyway, this may be a bug with the server but again, it's beyond my skill set.

In the end I ditched the CCHD and spent $20 of my hard earned cash at a local Walmart for the Onn 4K streaming box. No extra ads, no extra tracking that I can see, and runs a stock Android TV experience. 

And of course, it can actually handle 4K downscaling in-device. 


Regarding the bounty theory, generally that's poor taste for open source projects like this, but I don't know the feelings of the team here.

**Comment by nielsvanvelzen** [comment]
> I did make a PR (since accepted) to add the Chromecast HD (CCHD) to the list of devices which is forced to a max of 1080p, but it's not in any of the app releases yet. However the server appears to be ignoring the limit and sends the CCHD 4k anyway, this may be a bug with the server but again, it's beyond my skill set.

The pull request (#3820) is released in 0.17.1. The 1080p conditions require Jellyfin 10.9.10 to work properly as there was also an server-side issue (since 10.8.0) we discovered.

**Comment by majmongoose** [comment]
Aha! Thanks for the update Neils.

**Comment by BobbyWibowo** [comment]
Is this known to still be broken with server 10.9.11? Mine is still direct streaming 4k content, unfortunately.

I'm on a Chromecast HD, which supposedly is already hardcoded to require 1080p max, as of Jellyfin TV 0.17.5.

**Comment by andreasgerner** [comment]
@nielsvanvelzen are there any plans on continuing this topic? It is really annoying and a major downside of using Jellyfin for streaming to an AVR that setting preferred codecs (and disabling AAC) is not possible in any way if Android TV is used as a steaming device.

Edit: I also think this topic needs more attention as it may affect lots of users and keeps the Android TV app behind the WebUI or other platforms.

**Pull request closed without merging**
